massa-bootstrap/src/establisher.rs

```diff
@@ -1,48 +1,61 @@
 // Copyright (c) 2022 MASSA LABS <info@massa.net>
-use async_trait::async_trait;
 use massa_time::MassaTime;
-use std::{io, net::SocketAddr};
+use std::{
+    io,
+    net::{SocketAddr, TcpListener, TcpStream},
+};
 
 /// duplex connection
-pub type Duplex = tokio::net::TcpStream;
+pub trait Duplex:
+// static because need to send between threads :(
+    'static  + Send + tokio::io::AsyncReadExt + tokio::io::AsyncWriteExt + std::marker::Unpin
+{
+}
 
-/// Listener used to establish a Duplex
-pub type DuplexListener = tokio::net::TcpListener;
+impl Duplex for tokio::net::TcpStream {}
 
-#[async_trait]
 /// Specifies a common interface that can be used by standard, or mockers
 pub trait BSListener {
-    async fn accept(&mut self) -> io::Result<(Duplex, SocketAddr)>;
+    fn accept(&mut self) -> io::Result<(TcpStream, SocketAddr)>;
 }
 
-#[async_trait]
 /// Specifies a common interface that can be used by standard, or mockers
 pub trait BSConnector {
-    async fn connect(&mut self, addr: SocketAddr) -> io::Result<Duplex>;
-}
-
-/// Specifies a common interface that can be used by standard, or mockers
-pub trait BSEstablisher {
-    // TODO: this is needed for thread spawning. Once the listener is on-thread, the static
-    // lifetime can be thrown away.
-    // TODO: use super-advanced lifetime/GAT/other shenanigans to
-    // make the listener compatable with being moved into a thread
-    type Listener: BSListener + Send + 'static;
-    type Connector: BSConnector;
-    fn get_listener(&mut self, addr: SocketAddr) -> io::Result<Self::Listener>;
-    fn get_connector(&mut self, timeout_duration: MassaTime) -> io::Result<Self::Connector>;
+    fn connect(&mut self, addr: SocketAddr) -> io::Result<TcpStream>;
 }
 
 /// The listener we are using
 #[derive(Debug)]
-pub struct DefaultListener(DuplexListener);
+pub struct DefaultListener(TcpListener);
+impl DefaultListener {
+    /// Provides a standard TcpListener
+    pub fn new(addr: &SocketAddr) -> io::Result<DefaultListener> {
+        // Create a socket2 TCP listener to manually set the IPV6_V6ONLY flag
+        // This is needed to get the same behavior on all OS
+        // However, if IPv6 is disabled system-wide, you may need to bind to an IPv4 address instead.
+        let domain = match addr.is_ipv4() {
+            true => socket2::Domain::IPV4,
+            _ => socket2::Domain::IPV6,
+        };
+
+        let socket = socket2::Socket::new(domain, socket2::Type::STREAM, None)?;
+
+        if addr.is_ipv6() {
+            socket.set_only_v6(false)?;
```

Comment:
sure ?

Reply:
...hmmm. probably not, now that you mention it. At some point I removed the Establisher trait/struct entirely, and some wires must've gotten crossed in my head.

The fix will be to have this either in the establisher OR here, but not both.

Reply:
ok, same at line 103

Reply:
done: [0721d10](https://github.com/massalabs/massa/pull/3685/commits/0721d10d0ec91df983bce12e059cfcc8d62842c9)

Reply:
note: the setting to non-blocking is deferred until the `run_listener` method